torchmultimodal/modules/layers/attention.py

```diff
@@ -224,17 +217,14 @@ def forward(
         q: Tensor,
         k: Tensor,
         v: Tensor,
-        decode_step: Optional[int] = None,
-        decode_idx: Optional[Iterable[int]] = None,
     ) -> Tensor:
         mask = torch.Tensor(self.mask) if self.causal else None
```

Comment:
The mask is already created for you when you're decoding. what about the case when a user specifies an attention mask during encoding? I am thinking of making `FullAttention` the base self-attention class for all models. But in FLAVA, ALBEF there is attention masking used in the encoders. how do we reconcile this?

Reply:
By "mask created while decoding" Are you referring to 

https://github.com/facebookresearch/multimodal/blob/e274207e5c091d681bcb790d561fca0764148549/torchmultimodal/modules/layers/attention.py#L231-L232

?

I think mask should be created by a top level module (either vanilla transformer or that in the decoding mode) and passed in to the attention modules instead of being created inside the attention module which is not only inflexible but incurring the additional argument `shape`.

Is this consistent with "user specifies an attention mask during encoding"?

Reply:
Yes I agree with that, I think it's better design and it's consistent with other models. FLAVA passes in the attention mask at the highest level at `FLAVAModel`'s forward, for example. I can move the mask out of the attention modules in a separate PR for unifying `SelfAttention` and `MultiHeadAttention`

Reply:
Feel free to move the mask out of the attention module. I was thinking about doing that previously as the transformer decoder layers will depend on that.